flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/LegacyUserDefinedFunctionInference.java

```diff
@@ -0,0 +1,529 @@
+
+import org.apache.flink.shaded.guava31.com.google.common.primitives.Primitives;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.math.BigDecimal;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static org.apache.flink.table.types.logical.utils.LogicalTypeUtils.toInternalConversionClass;
+import static org.apache.flink.table.types.utils.TypeConversions.fromLegacyInfoToDataType;
+
+/**
+ * Ported {@code UserDefinedFunctionUtils} to run some of the type inference for legacy functions in
+ * Table API.
+ */
+@Internal
+@Deprecated
+public class LegacyUserDefinedFunctionInference {
+
+    public static InputTypeStrategy getInputTypeStrategy(ImperativeAggregateFunction<?, ?> func) {
+        return new InputTypeStrategy() {
+            @Override
+            public ArgumentCount getArgumentCount() {
+                return ConstantArgumentCount.any();
+            }
+
+            @Override
+            public Optional<List<DataType>> inferInputTypes(
+                    CallContext callContext, boolean throwOnFailure) {
+                final List<DataType> argumentDataTypes = callContext.getArgumentDataTypes();
+                final DataType accType = getAccumulatorType(func);
+                final LogicalType[] input =
+                        Stream.concat(Stream.of(accType), argumentDataTypes.stream())
+                                .map(DataType::getLogicalType)
+                                .toArray(LogicalType[]::new);
+
+                final Optional<Optional<Method>> foundMethod =
+                        Stream.of(
+                                        logicalTypesToInternalClasses(input),
+                                        logicalTypesToExternalClasses(input))
+                                .map(
+                                        signature ->
+                                                getUserDefinedMethod(
+                                                        func,
+                                                        "accumulate",
+                                                        signature,
+                                                        input,
+                                                        cls ->
+                                                                Stream.concat(
+                                                                                Stream.of(accType),
+                                                                                Arrays.stream(cls)
+                                                                                        .skip(1)
```

Comment:
Ok, I have to ask... why are we skipping the first class?

It'd be lovely if there's any way to make this more clear.  (I don't have suggestions, and I imagine you have tried.)

Reply:
Because of 

> A no-goal of the PR is to make the logic in LegacyUserDefinedFunctionInference more maintainable.

and

https://github.com/apache/flink/blob/977463cce3ea0f88e2f184c30720bf4e8e97fd4a/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/functions/utils/UserDefinedFunctionUtils.scala#L256